devise を使ってユーザー認証を実装する

gaki273 opened the pull request.

itagakisan committed:
deviseを導入、インストール

itagakisan committed:
Userモデルを作成

itagakisan committed:
Deviseの認識ビューファイル、コントローラファイルを追加

itagakisan committed:
アカウント作成後、自動ログインし本一覧画面に遷移するように修正

itagakisan committed:
ユーザーのプロフィール画面作成、ルーティングを設定

itagakisan committed:
ユーザーのプロフィール画面からの画面遷移を追加

itagakisan committed:
ユーザーのプロフィール画面から本の一覧画面への遷移を追加

itagakisan committed:
プロフィールをパスワードなしで更新するように変更

itagakisan committed:
プロフィール更新後。プロフィール画面への遷移を追加

itagakisan committed:
ユーザープロフィールに郵便番号・住所・自己紹介文の入力・表示欄を追加、strong parametersを追加

itagakisan committed:
ユーザーの一覧ページを追加

itagakisan committed:
ユーザーの詳細画面とマイページを分離

itagakisan committed:
configure_permitted_parametersメソッドを修正

itagakisan committed:
ユーザー一覧にてログインしているユーザーを表示しないように処理を追加

itagakisan committed:
ログアウト機能を追加

itagakisan committed:
アプリの利用にログインを必須とするように処理を追加

itagakisan committed:
letter_opener_webを導入、パスワードリセットメール確認機能を追加

itagakisan committed:
ユーザー各画面のナビゲーションを改善

itagakisan committed:
devise-i18nの導入、deviseの日本語ロケールファイルを追加

itagakisan committed:
アカウント登録画面の日本語対応

itagakisan committed:
パスワードの再設定送信画面の日本語対応

itagakisan committed:
パスワードの再設定画面の日本語対応

itagakisan committed:
ユーザープロフィール画面の日本語対応

itagakisan committed:
プロフィール編集画面の日本語対応

itagakisan committed:
ユーザー一覧画面を修正

itagakisan committed:
メソッドの引数とパラメータを修正

itagakisan committed:
フラッシュメッセージを表示するように処理を追加

itagakisan committed:
ログイン画面の日本語対応

itagakisan committed:
deviseの翻訳ファイルに追記

itagakisan committed:
不要なファイルを削除

itagakisan committed:
本の一覧画面の日本語対応

itagakisan committed:
ユーザーの一覧画面の日本語対応

itagakisan committed:
ユーザー一覧画面にページネーション処理を追加

itagakisan committed:
deviseルートにsessionsコントローラーを追加

itagakisan committed:
アカウント削除の確認ダイアログの日本語対応

maedana commented on Gemfile:
```diff
@@ -83,3 +84,7 @@ end
+
+gem 'devise'
+
+gem 'devise-i18n', '~> 1.14'
```
これに限らずですが、特別な理由がなければバージョン指定なしで最新版を使いましょうー。Railsのように勝手にバージョンが上がっては困るものだけ固定したらよいかと

gaki273 replied:
かしこまりました！
修正いたします。

maedana commented on app/controllers/users/passwords_controller.rb:
```diff
@@ -0,0 +1,34 @@
+  # def after_sending_reset_password_instructions_path_for(resource_name)
+  #   super(resource_name)
+  # end
+end
```
中身が空のものを配置する意味がないかなーと思います。この辺読んで適切な対応をしましょうー。 https://github.com/heartcombo/devise?tab=readme-ov-file#configuring-controllers

gaki273 replied:
かしこまりました。
適切な対応を行い、修正いたします。

maedana commented on app/models/user.rb:
```diff
@@ -0,0 +1,19 @@
+    result = update(params)
+    clean_up_passwords
+    result
+  end
```
本体に似たようなものがあるんですけど何故再実装する必要があるのか説明してみて頂けますか?

https://github.com/heartcombo/devise/blob/cf93de390a29654620fdf7ac07b4794eb95171d0/lib/devise/models/database_authenticatable.rb#L120

gaki273 replied:
申し訳ございません。すでに本体に実装されていたのですね。確認が足りていませんでした。
今後は、使用するものについて確認するように気を付けてまいります。
削除しても同様の動作が可能なことを確認しましたので、削除して修正いたします。

maedana commented on config/locales/devise.views.ja.yml:
```diff
@@ -0,0 +1,155 @@
+      not_locked: はロックされていません。
+      not_saved:
+        one: "%{count} 件のエラーが発生したため %{resource} は保存されませんでした。"
+        other: "%{count} 件のエラーが発生したため %{resource} は保存されませんでした。"
```
devise-i18nが入ってますし、特にカスタマイズしてないならこのファイルなくても動くのではないですかね? なぜ必要なんでしょう?

gaki273 replied:
そのファイルに足りない日本語訳について追記していく形だと考えていましたが、足りないものについては`translation_ja.yml`のほうに追記していけばよろしいでしょうか。
対象のファイルは削除し、足りない日本語訳を`translation_ja.yml`に追記していく形で修正いたします。

maedana commented on config/routes.rb:
```diff
@@ -1,5 +1,14 @@
+  devise_for :users, controllers: {
+    registrations: 'users/registrations',
+    passwords: 'users/passwords',
+    sessions: 'users/sessions'
```
標準と動きを変えたいのは`registrations`だけではないですかね?

gaki273 replied:
`passwords`と`sessions`は、Turbo使用時のフラッシュメッセージ表示問題を回避するため設定していましたが、各カスタムコントローラに必要な処理とTurbo対応を直接実装する方式に変更いたします。

maedana replied:
確認してみましたが、Flashメッセージ表示の問題はcontrollerではなくview側の問題なのではないですか? 単にviewの配置の問題だと思いますよ。デフォルトのcontrollerがカスタムしたviewを参照してくれるようにさえなればよいのではないかと。以下根拠です。

デフォルトのcontrollerでは
```
Started GET "/users/sign_in" for 127.0.0.1 at 2025-07-08 14:48:30 +0900
Processing by Devise::SessionsController#new as TURBO_STREAM
  Rendering layout layouts/application.html.erb
  Rendering /home/maedana/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/devise-i18n-1.14.0/app/views/devise/sessions/new.html.erb within layouts/application
  Rendered /home/maedana/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/devise-i18n-1.14.0/app/views/devise/shared/_links.html.erb (Duration: 2.3ms | Allocations: 592)
  Rendered /home/maedana/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/devise-i18n-1.14.0/app/views/devise/sessions/new.html.erb within layouts/application (Duration: 9.7ms | Allocations: 2135)
  Rendered layout layouts/application.html.erb (Duration: 18.5ms | Allocations: 4378)
Completed 200 OK in 24ms (Views: 19.9ms | ActiveRecord: 0.0ms | Allocations: 5730)
```

カスタムcontrollerでは
```
Started GET "/users/sign_in" for 127.0.0.1 at 2025-07-08 14:45:44 +0900
Processing by Users::SessionsController#new as TURBO_STREAM
  Rendering layout layouts/application.html.erb
  Rendering users/sessions/new.html.erb within layouts/application
  Rendered users/shared/_links.html.erb (Duration: 1.7ms | Allocations: 179)
  Rendered users/sessions/new.html.erb within layouts/application (Duration: 13.6ms | Allocations: 1692)
  Rendered layout layouts/application.html.erb (Duration: 33.7ms | Allocations: 3946)
Completed 200 OK in 39ms (Views: 36.7ms | ActiveRecord: 0.0ms | Allocations: 5305)
```

となります。sessions/newが前者では、`devise-i18n-1.14.0/app/views/devise/sessions/new.html.erb`で、後者ではリポジトリ内の`users/sessions/new.html.erb`なわけです。

https://github.com/heartcombo/devise?tab=readme-ov-file#configuring-views に

> After doing so, you will be able to have views based on the role like users/sessions/new and admins/sessions/new. If no view is found within the scope, Devise will use the default view at devise/sessions/new. You can also use the generator to generate scoped views:

とあるように、`default view at devise/sessions/new`なわけです。以上踏まえると、sessions/passwordsについてはやはりcontroller不要かと思いますし、現在の実装もちょっと変ですね。

```ruby
class Users::SessionsController < Devise::SessionsController
  def respond_to_on_destroy
    redirect_to new_user_session_path
  end
end
```
となってますが、オリジナルの実装が
```ruby
  def respond_to_on_destroy
    # We actually need to hardcode this as Rails default responder doesn't
    # support returning empty response on GET request
    respond_to do |format|
      format.all { head :no_content }
      format.any(*navigational_formats) { redirect_to after_sign_out_path_for(resource_name), status: Devise.responder.redirect_status }
    end
  end
```
で、最終的に`redirect_to after_sign_out_path_for(resource_name)`しており、`ApplicationController`で
```ruby
  def after_sign_out_path_for(_resource_or_scope)
    new_user_session_path
  end
```
してますから、結局最終的なリダイレクト先は同じです。passwordsについてもオリジナルが
```ruby
    def after_sending_reset_password_instructions_path_for(resource_name)
      new_session_path(resource_name) if is_navigational_format?
    end
```
ですから同様に無意味です。結局は単にViewの問題でしかないかなと。以上いかがでしょうか。

gaki273 replied:
とても丁寧なご指摘ありがとうございます。
自分自身も修正していくうちに、View側の問題もあるかなと感じることがありました。
いただいたコメントをもとに、Viewのほうを重点的に見直し、修正いたします。

gaki273 replied:
>デフォルトのcontrollerがカスタムしたviewを参照してくれるようにさえなればよいのではないかと。

View側の問題というよりも、上記のご指摘の通りカスタムしたViewを参照できていないことが原因でした。
`config/initializers/devise.r`で、`config.scoped_views = true`を追加したら、デフォルトのcontrollerがカスタムしたviewを参照するようになりました！
不要なファイルとコントローラ指定を削除し修正いたしました。

maedana replied:
なるほどー、想定していたのは`users`配下にあるdevies用のviewを`devise`配下に配置することでしたが、確かに`config.scoped_views = true`でいいですね :+1:

itagakisan committed:
不要な記述とメソッドを削除

itagakisan committed:
Deviseコントローラーのリダイレクト処理をカスタマイズ 修正

[目的]
- Turbo環境でのフラッシュメッセージ表示を改善

[変更内容]
- ログアウト・パスワードリセット後のリダイレクト先を指定

itagakisan committed:
deviseのロケールファイルを削除し必要な翻訳をtranslation.ymlに追加

maedana commented on config/locales/translation_ja.yml:
```diff
@@ -2,10 +2,36 @@ ja:
+    passwords:
+      edit:
+        minimum_password_length:
+          other: "（%{count}字以上）"
```
https://github.com/tigrish/devise-i18n/blob/master/rails/locales/ja.yml#L127 でカバーされてたりしないですかね? 書かないと動作しませんでした?

gaki273 replied:
各View（passwords、registrations等）で使用する翻訳は、それぞれ対応するセクションごとに用意しないといけないものと勘違いしておりました。
実際には共通で使用される要素についてはsharedセクションにまとめて定義されているのですね。
devise-i18nの構造について理解が足りておりませんでした。

共通の翻訳キーに統一して修正いたします。

itagakisan committed:
不要な翻訳を削除、i18nキーを共通化 リファクタリング

[目的]
- deviseのi18n設定を共通化し、重複を排除するため

[変更内容]
- パスワードの最小文字数に関するi18nキーを共通化
- 不要になった翻訳定義を削除

itagakisan committed:
不要なファイル、コントローラ指定を削除、カスタムしたViewファイルを読み込むように設定 リファクタリング

[目的]
- deviseのコントローラーをデフォルト実装に戻し、冗長なコードを削除する

[変更内容]
- 不要なUsers::PasswordsControllerとUsers::SessionsControllerを削除
- config.scoped_views = true を設定して、カスタムビューが読み込まれるようにする
- routes.rbから不要なコントローラー指定を削除

itagakisan committed:
不要なファイルを削除

gaki273 commented:
見直していたところ、特にカスタムしていないViewファイルについては削除を行いました。
